src/main/scala/latis/input/HapiBinaryAdapter.scala

```diff
@@ -0,0 +1,128 @@
+
+  /**
+   * Provides a Stream of records as Arrays of bytes.
+   */
+  def recordStream(uri: URI): Stream[IO, Array[Byte]] =
+    StreamSource.getStream(uri)
+      .chunkN(blockSize)
+      .map(_.toArray)
+  
+  /**
+   * Parses a record into a Sample. 
+   * Returns None if the record is invalid.
+   */
+  def parseRecord(record: Array[Byte]): Option[Sample] = {
+    // We assume one value for each scalar in the model.
+    // Note that Samples don't capture nested tuple structure.
+    // Assume uncurried model (no nested function), for now.
+    /*
+     * TODO: Traverse the model to build nested functions (Function in range) 
+     * should we apply currying logic or require operation? 
+     * curry should be cheap for ordered data
+     */
+
+    // Get Vectors of domain and range Scalar types.
+    val (dtypes, rtypes) = model match {
+      case Function(d, r) => (d.getScalars, r.getScalars)
+      case _ => (Vector.empty, model.getScalars)
+    }
+
+    // Extract the data values from the record
+    // and split into Vectors of domain and range.
+    val data = extractData(record)
+    val (ds, rs) = data.splitAt(dtypes.length)
+
+    // Construct a Sample from the domain and range values.
+    if (rtypes.length != rs.length) None //invalid record
+    else 
+      Some(Sample(ds, rs))
+  }
+
+  /**
+   * Extracts the data values from the given record as a List.
+   */
+  private def extractData(record: Array[Byte]): List[Data] = {
+    val buffer = ByteBuffer.wrap(record).order(order)
+    
+    model.getScalars.map { s => 
+      s("type") match {
+        case Some("short")  => Data(buffer.getDouble.toShort)
+        case Some("int")    => Data(buffer.getDouble.toInt)
+        case Some("long")   => Data(buffer.getDouble.toLong)
+        case Some("float")  => Data(buffer.getDouble.toFloat)
+        case Some("double") => Data(buffer.getDouble)
+        case Some("string") => 
+          val length = bytesToRead(s)
+          val bytes = new Array[Byte](length)
+          buffer.get(bytes, 0, length)
+          Data(new String(bytes, StandardCharsets.UTF_8))
+        case Some(_) => ??? //unsupported type
+        case None => ??? //type not defined
+      }
+    }
+  }
+
+  /**
+   * Gets the number of bytes to read for the given scalar
+   * according to the HAPI specification.
+   * 
+   * Note that some values deviate from the number of bytes
+   * needed to represent the corresponding primitive type
+   * because four byte and floating point values are always 
+   * IEEE 754 double precision values.
+   * 
+   * Parameters of type string and isotime have a "length" 
+   * specified in the info header that indicates how many bytes 
+   * to read for each string value.
+   */
+  private def bytesToRead(s: Scalar): Int = {
+    s("type") match {
+      //case Some("char")   => 2
+      case Some("short")  => 2
+      case Some("int")    => 8 //not 4
+      case Some("long")   => 8
```

Comment:
Maybe 4 for the integer types.

Reply:
Will you please make this a requested change so I don't forget to address it. I'm not sure what the right answer is at the moment.

Reply:
Never mind, code has been updated to read 4 bytes for integers.